src/antidote_crdt_secure_counter_pn.erl

```diff
@@ -0,0 +1,173 @@
+%% -------------------------------------------------------------------
+%% Copyright <2020> <
+%%  Technische Universität Kaiserslautern, Germany
+%%  Université Pierre et Marie Curie / Sorbonne-Université, France
+%%  Universidade NOVA de Lisboa, Portugal
+%%  Université catholique de Louvain (UCL), Belgique
+%%  INESC TEC, Portugal
+%% >
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either expressed or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% List of the contributors to the development of Antidote: see AUTHORS file.
+%% Description and complete License: see LICENSE file.
+%% -------------------------------------------------------------------
+
+%% antidote_crdt_secure_counter_pn: A convergent, replicated, operation
+%% based secure (using the Paillier cryptosystem) PN-Counter.
+
+-module(antidote_crdt_secure_counter_pn).
+
+-behaviour(antidote_crdt).
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-export([
+    new/0,
+    value/1,
+    downstream/2,
+    update/2,
+    equal/2,
+    to_binary/1,
+    from_binary/1,
+    is_operation/1,
+    require_state_downstream/1
+]).
+
+-type freshness() :: fresh | spoiled.
+-type delta() :: integer().
+-type nsquare() :: integer().
+-type state() :: {freshness(), integer(), nsquare()}.
+-type op() :: {increment, delta()} | {increment, {delta(), nsquare()}}.
+-type effect() :: delta() | {delta(), nsquare()}.
+
+%% @doc Create a new, empty 'antidote_crdt_secure_counter_pn'
+-spec new() -> state().
+new() ->
+    {fresh, 0, 0}.
```

Comment:
If the initial N² value is 0, this could lead to a division by 0 error in the update function later. Maybe it is better to always require N² in the operations?

Reply:
> Maybe it is better to always require N² in the operations?
Yes, I think this is a good idea.